R/tolerance.cca.R

```diff
@@ -84,11 +89,14 @@ tolerance.cca <- function(x, choices = 1:2,
         if(doN2) {
             y <- sweep(Y, 2, Ytot, "/")^2
             N2 <- 1 / colSums(y, na.rm = TRUE) # 1/H
+            ## avoid almost-1 for species present only once
+            N2 <- zapsmall(N2 - 1L) + 1L
             ## avoid "negative zeros" form 1 - 1/N2 when N2 ~ 1
             res <- sweep(res, 1, sqrt(pmax(1 - 1/N2, 0)), "/")
         }
     }
     res[!is.finite(res)] <- 0 # some values can be Inf or NaN but are really 0
+    res <- zapsmall(res)      # almost-zero tolerances should be zero
```

Comment:
this cannot be zapped because the non-zero bits are the first (and only) ones, and you need
res[res < sqrt(.Machine$double.eps)] <- 0 or similar.